pkg/csi/service/node.go

```diff
@@ -674,109 +683,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
 		log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
 		return nodeInfoResponse, nil
 	}
-	var cfg *cnsconfig.Config
-	cfgPath = os.Getenv(cnsconfig.EnvVSphereCSIConfig)
-	if cfgPath == "" {
-		cfgPath = cnsconfig.DefaultCloudConfigPath
-	}
-	cfg, err := cnsconfig.GetCnsconfig(ctx, cfgPath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
-			nodeInfoResponse = &csi.NodeGetInfoResponse{
-				NodeId:            nodeID,
-				MaxVolumesPerNode: maxVolumesPerNode,
+
+	var (
+		accessibleTopology map[string]string
+		err error
+	)
+	if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ImprovedVolumeTopology) {
+		accessibleTopology, err = fetchTopologyLabelsUsingCR(ctx, nodeName, nodeID)
+	} else {
+		var cfg *cnsconfig.Config
+		cfgPath = os.Getenv(cnsconfig.EnvVSphereCSIConfig)
+		if cfgPath == "" {
+			cfgPath = cnsconfig.DefaultCloudConfigPath
+		}
+		cfg, err := cnsconfig.GetCnsconfig(ctx, cfgPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
+				nodeInfoResponse = &csi.NodeGetInfoResponse{
+					NodeId:            nodeID,
+					MaxVolumesPerNode: maxVolumesPerNode,
+				}
+				log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
+				return nodeInfoResponse, nil
 			}
-			log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
-			return nodeInfoResponse, nil
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to read CNS config. Error: %v", err)
 		}
-		return nil, logger.LogNewErrorCodef(log, codes.Internal,
-			"failed to read cnsconfig. err: %v", err)
+		accessibleTopology, err = fetchTopologyLabelsUsingVCCreds(ctx, nodeID, cfg)
+	}
+	if err != nil {
+		return nil, err
 	}
-	var accessibleTopology map[string]string
+
 	topology := &csi.Topology{}
+	if len(accessibleTopology) > 0 {
+		topology.Segments = accessibleTopology
+	}
+	nodeInfoResponse = &csi.NodeGetInfoResponse{
+		NodeId:             nodeID,
+		MaxVolumesPerNode:  maxVolumesPerNode,
+		AccessibleTopology: topology,
+	}
+	log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
```

Comment:
How often is this function called? Is there value in logging the same info everytime?

Reply:
This function is called just once per node. We generally prefer to log the response of every API before returning, so followed it here too.